tiled/_tests/test_structured_array.py

```diff
@@ -33,7 +34,6 @@ def test_read():
         dtype=[("name", "U10"), ("age", "i4"), ("weight", "f4")],
     )
     tree = MapAdapter({"A": ArrayAdapter.from_array(data)})
-
-    client = from_tree(tree)
-
+    with Context.from_app(build_app(tree)) as context:
```

Comment:
This pattern is used a lot, and this is somewhat picky, but could you make it a little more DRY but defining a function to be called each time:

```python
def client_app(tree):
    with Context.from_app(tree) as context:
        client = from_context(context)
    return client
```

Then each test that obtains a client could:

```python
client = cleint_app(tree)
```

This wouldn't save all that much code, but if your be a nice future proofing if you realize that your strategy needs to change slightly, it could be changed in one place.

Reply:
That doesn't quite work because:

```py
def client_app(tree):
    with Context.from_app(build_app(tree)) as context:
        client = from_context(context)
    # Server event loop stops here.
    return client  # dead on arrival
```

It could be done like:

```py
import contextlib

@contextmanager
def client_app(tree):
    with Context.from_app(build_app(tree)) as context:
        client = from_context(context)
        yield client
```

If it were always that simple, I think I would agree we want a function to stay DRY. But it gets complicated.

* We sometimes build the app based on `config` instead of a `tree`, so we'd need a `client_app_from_config` variant. Stepping back, this is exactly what `from_tree` and `from_config` were (minus the context manager aspect).
* in many of the tests we need to pass in additional parameters to either `Context.from_app` or `from_context` or both, such as
  ```py
  with Context.from_app(tree, authentication={...}) as context:
      client = from_context(context, structure_clients="dask")
  ```
  so `client_app` would have to handle a mixture of server-related and client-related parameters---kind of confusing. The `from_tree` and `from_config` functions also had this downside.

Is it better to be DRY with `Context.from_app`, `from_context`, `from_tree`, and `from_config` in the mix, or is it better to be repetitive with just `Context.from_app` and `from_context`?

For test modules that need a consistent `tree` or `config` across many tests, I did the equivalent of `client_app` as a `pytest.fixture`:

```py
@pytest.fixture(scope="module")
def client():
    app = build_app(tree)
    with Context.from_app(app) as context:
        client = from_context(context)
        yield client
```

such as in `test_search.py`.